File: Study Plan/Top Interview 150/0020 Valid Parentheses/ValidParentheses.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    20. Valid Parentheses
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/valid-parentheses/solutions/4422923/c-based-on-hints-beats-92-in-runtime-time-and-space-complexity-o-n/

    - Runtime 49 ms (Beats 92.18% of users with C#)
    - Memory 39.50 MB (Beats 25.21% of users with C#)
    
*/

public class Solution
{
    public bool IsValid(string s)
    {
        if (s.Length == 1) return false;

        // Hint 1
        // Use a stack of characters.
        var characters = new Stack<char>(s.Length);

        for (var character = 0; character < s.Length; character++)
        {
            // Hint 2
            // When you encounter an opening bracket,
            if (s[character] == '('
                || s[character] == '['
                || s[character] == '{')
            {
                // push it to the top of the stack.
                characters.Push(s[character]);
            } 
            // Hint 3
            // When you encounter a closing bracket, check if the top of the stack
            // was the opening for it.
            else if ((s[character] == ')' && characters.Any() && characters.Peek() == '(')
                || (s[character] == ']' && characters.Any() && characters.Peek() == '[')
                || (s[character] == '}' && characters.Any() && characters.Peek() == '{'))
            {
                // If yes, pop it from the stack.
                characters.Pop();
            }
            else
            {
                // Otherwise, return false.
                return false;
            }
        }

        return !characters.Any();
    }
}

File: Study Plan/Top Interview 150/0020 Valid Parentheses/ValidParentheses.java
/*  
    Problem
    --------------------------------------------------------------------------------
    20. Valid Parentheses
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/valid-parentheses/solutions/4433109/java-based-on-hints-beats-82-in-runtime-time-and-space-complexity-o-n/

    - Runtime 2 ms (Beats 67.06% of users with Java)
    - Memory 41.34 MB (Beats 50.86% of users with Java)
*/

class Solution {
    public boolean isValid(String s) {
        if (s.length() == 1) return false;
        // Hint 1
        // Use a stack of characters.
        // https://www.baeldung.com/java-deque-vs-stack
        Deque<Character> characters = new ArrayDeque<>(s.length());
        for (int character = 0; character < s.length(); character++) {
            // Hint 2
            // When you encounter an opening bracket,
            if (s.charAt(character) == '('
                || s.charAt(character) == '['
                || s.charAt(character) == '{') {
                // push it to the top of the stack.
                characters.push(s.charAt(character));
            // Hint 3
            // When you encounter a closing bracket, check if the top of the stack
            // was the opening for it.
            }  else if ((s.charAt(character) == ')' 
                && !characters.isEmpty() 
                && characters.peek() == '(')
                || 
                (s.charAt(character) == ']' 
                && !characters.isEmpty() 
                && characters.peek() == '[')
                || 
                (s.charAt(character) == '}' 
                && !characters.isEmpty() 
                && characters.peek() == '{')) {
                // If yes, pop it from the stack.
                characters.pop();
            } else {
                // Otherwise, return false.
                return false;
            }
        }
        return characters.isEmpty();
    }
}
